include/RAJA/index/IndexSet.hpp

```diff
@@ -433,14 +433,14 @@ class TypedIndexSet<T0, TREST...> : public TypedIndexSet<TREST...>
   /// This TypedIndexSet will not change and the created "slice" into it
   /// will not own any of its segments.
   ///
-  TypedIndexSet<T0, TREST...> *createSlice(int begin, int end)
+  TypedIndexSet<T0, TREST...> createSlice(int begin, int end)
   {
-    TypedIndexSet<T0, TREST...> *retVal = new TypedIndexSet<T0, TREST...>();
+    TypedIndexSet<T0, TREST...> retVal;
 
     int minSeg = RAJA::operators::maximum<int>{}(0, begin);
     int maxSeg = RAJA::operators::minimum<int>{}(end, getNumSegments());
     for (int i = minSeg; i < maxSeg; ++i) {
-      segment_push_into(i, *retVal, PUSH_BACK, PUSH_NOCOPY);
+      segment_push_into(i, retVal, PUSH_BACK, PUSH_NOCOPY);
```

Comment:
Maybe this is just a naming thing, but is this right? It reads to me like the call is being asked to incorporate the indexset without copying it, or taking ownership of it, with the `PUSH_NOCOPY`, then the indexset gets deconstructed at the end of the function?

Reply:
I believe the intent here is to create a new IndexSet containing a subset of Segments from the original IndexSet specified by the begin/end Segment indices (i.e., a "slice" of the original. The Segments (if they are ListSegments) don't need to be copied to the slice version, because the assumption is that the lifetime of the original contains the lifetime of the slice version.  The object that the function returns should be a copy of the slice IndexSet created here. This works with our unit tests that check various ways to construct an index set via different means to add segments. However, looking over the tests, it seems that we can do better -- the test assumes that the orginal index set and what gets returned by this method are the same; i.e., number of segments, segments are equivalent, etc.  I will add a JIRA issue to beef up the tests. Does this answer your concerns?

Reply:
Gotcha, yes that makes sense.  As long as the lifetime of the return is strictly shorter than the source that should be fine, somehow I was reading it as less bounded than that.